File: apps/client/src/data/skills.ts
/**
 * Efeito das skills sobre os números mostrados no hangar.
 *
 * Espelho de `crates/sim-core/src/ship/skills.rs`, e pelo mesmo motivo
 * do espelho de armas: o cliente precisa ANTECIPAR o que o servidor vai
 * calcular, para o painel do estaleiro não mentir.
 *
 * O painel mostrava o DPS do equipamento puro. Com o ramo de combate
 * comprado, o dano real era 20,75% maior que o exibido — o jogador
 * gastava pontos e o número não se mexia, exatamente o sintoma que fazia
 * a árvore parecer decorativa.
 *
 * Quem DECIDE continua sendo o servidor. Isto é previsão, não autoridade:
 * o cliente manda ids e recebe o resultado.
 */

export interface CombatMods {
  damageMult: number;
  fireRateMult: number;
  /** 0..1 — fração do dano que ignora o escudo alvo. */
  shieldPierce: number;
  chargeTimeMult: number;
}

export const NO_MODS: CombatMods = {
  damageMult: 1,
  fireRateMult: 1,
  shieldPierce: 0,
  chargeTimeMult: 1,
};

/**
 * Efeito de um nó. `undefined` para nós que não tocam no combate.
 *
 * Os valores batem com `node_effect` em Rust — há teste de paridade
 * contra uma fixture gerada pelo próprio servidor.
 */
const EFEITOS: Record<string, Partial<CombatMods>> = {
  combat_t1: { damageMult: 1.05 },
  combat_t2: { fireRateMult: 1.1 },
  // "Critical Strike +15% crit chance" vira o valor ESPERADO
  // equivalente: sem sistema de crítico, um acerto aleatório seria dano
  // não determinístico, e o servidor precisa ser replicável.
  combat_t3: { damageMult: 1.15 },
  combat_t4: { shieldPierce: 0.1 },
  combat_t5: { damageMult: 1.12 },
};

/**
 * Acumula os nós desbloqueados.
 *
 * Multiplicadores compõem por MULTIPLICAÇÃO (dois nós de +10% dão +21%,
 * não +20%) e a perfuração soma com teto em 1 — as duas regras iguais às
 * do servidor. Divergir aqui faria o hangar prometer um número e a
 * arena entregar outro.
 */
export function combatMods(nodeIds: readonly string[]): CombatMods {
  const acc: CombatMods = { ...NO_MODS };
  for (const id of nodeIds) {
    const e = EFEITOS[id];
    if (!e) continue;
    acc.damageMult *= e.damageMult ?? 1;
    acc.fireRateMult *= e.fireRateMult ?? 1;
    acc.chargeTimeMult *= e.chargeTimeMult ?? 1;
    acc.shieldPierce = Math.min(1, acc.shieldPierce + (e.shieldPierce ?? 0));
  }
  return acc;
}

/** True se os modificadores mudam alguma coisa de fato. */
export function hasEffect(m: CombatMods): boolean {
  return (
    m.damageMult !== 1 || m.fireRateMult !== 1 || m.shieldPierce > 0 || m.chargeTimeMult !== 1
  );
}

/** Ids de nó com efeito de combate — para os testes de paridade. */
export function combatNodeIds(): string[] {
  return Object.keys(EFEITOS);
}

/** Só a parte de combate de `AggregateStats`, para não acoplar os módulos. */
interface StatsComCombate {
  damage: number;
  fireRate: number;
  dps: number;
}

/**
 * Aplica os modificadores aos números do painel.
 *
 * O `dps` é RECALCULADO de dano × cadência em vez de multiplicado pelo
 * ganho de dano: as duas skills se compõem, e multiplicar só o dano
 * perderia o ganho de cadência.
 */
export function applyModsToStats<T extends StatsComCombate>(stats: T, m: CombatMods): T {
  const damage = stats.damage * m.damageMult;
  const fireRate = stats.fireRate * m.fireRateMult;
  return {
    ...stats,
    damage: Math.round(damage * 10) / 10,
    fireRate: Math.round(fireRate * 100) / 100,
    dps: Math.round(damage * fireRate * 10) / 10,
  };
}

File: crates/sim-core/src/ship/skills.rs
//! Efeito das skills sobre o combate.
//!
//! A árvore de skills existia só como TEXTO: "+5% weapon damage",
//! "+10% fire rate", "10% dmg bypasses shield" apareciam na interface,
//! o jogador gastava pontos, e nada disso chegava à simulação. O
//! servidor nunca via as skills — o dano saía puro do catálogo de armas.
//!
//! Aqui os nós viram números, e é o SERVIDOR quem os aplica: o cliente
//! manda apenas os ids desbloqueados, exatamente como faz com o
//! loadout. Um cliente que mentisse um id inexistente não ganha nada,
//! porque ids desconhecidos são ignorados.
//!
//! Os ids batem com `apps/client/src/data/skillTree.json`.

use serde::{Deserialize, Serialize};

use super::weapons::WeaponProfile;

/// Modificadores acumulados de combate.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct CombatMods {
    /// Multiplicador de dano por projétil.
    pub damage_mult: f32,
    /// Multiplicador de cadência.
    pub fire_rate_mult: f32,
    /// Fração do dano que ignora o escudo e vai direto ao casco (0..1).
    pub shield_pierce: f32,
    /// Multiplicador do tempo de carga. < 1 carrega mais rápido.
    pub charge_time_mult: f32,
}

impl Default for CombatMods {
    fn default() -> Self {
        Self {
            damage_mult: 1.0,
            fire_rate_mult: 1.0,
            shield_pierce: 0.0,
            charge_time_mult: 1.0,
        }
    }
}

/// Efeito de um único nó da árvore.
///
/// Devolve `None` para nós que não tocam no combate (navegação,
/// engenharia): eles continuam existindo na árvore, só não somam aqui.
fn node_effect(node_id: &str) -> Option<CombatMods> {
    let m = match node_id {
        // "Sharpened Aim — +5% weapon damage"
        "combat_t1" => CombatMods {
            damage_mult: 1.05,
            ..Default::default()
        },
        // "Rapid Fire — +10% fire rate"
        "combat_t2" => CombatMods {
            fire_rate_mult: 1.10,
            ..Default::default()
        },
        // "Critical Strike — +15% crit chance"
        //
        // Sem sistema de crítico na simulação, um acerto crítico de 15%
        // seria dano invisível e não determinístico. Vira o valor
        // ESPERADO equivalente (15% de chance de +100% = +15% médio),
        // que é determinístico e replicável — requisito do servidor.
        "combat_t3" => CombatMods {
            damage_mult: 1.15,
            ..Default::default()
        },
        // "Armor Piercing — 10% dmg bypasses shield"
        "combat_t4" => CombatMods {
            shield_pierce: 0.10,
            ..Default::default()
        },
        // "Annihilator — +25% damage vs structures"
        //
        // Só naves e corpos celestes existem hoje; aplicado como dano
        // geral menor (+12%) para não virar um nó morto no fim do ramo.
        "combat_t5" => CombatMods {
            damage_mult: 1.12,
            ..Default::default()
        },
        _ => return None,
    };
    Some(m)
}

/// Soma os nós desbloqueados num único conjunto de modificadores.
///
/// Multiplicadores se compõem por MULTIPLICAÇÃO (dois nós de +10% dão
/// +21%, não +20%): somar deixaria o fim de um ramo longo desproporcional
/// e é a origem clássica de builds degeneradas.
///
/// A perfuração de escudo se acumula por soma, mas com teto em 1.0 —
/// ignorar o escudo inteiro já é o máximo que o efeito pode significar.
pub fn combat_mods(node_ids: &[String]) -> CombatMods {
    let mut acc = CombatMods::default();
    for id in node_ids {
        let Some(m) = node_effect(id) else { continue };
        acc.damage_mult *= m.damage_mult;
        acc.fire_rate_mult *= m.fire_rate_mult;
        acc.charge_time_mult *= m.charge_time_mult;
        acc.shield_pierce = (acc.shield_pierce + m.shield_pierce).min(1.0);
    }
    acc
}

/// Aplica os modificadores ao perfil da arma.
///
/// A `charge_time` menor não é só conveniência: ela muda o ritmo do
/// combate, então passa pelo mesmo caminho de todo o resto — o servidor.
pub fn apply_to_weapon(w: &WeaponProfile, m: &CombatMods) -> WeaponProfile {
    WeaponProfile {
        damage: w.damage * m.damage_mult,
        fire_rate: w.fire_rate * m.fire_rate_mult,
        charge_time: w.charge_time * m.charge_time_mult,
        ..*w
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ship::weapons::weapon_profile;

    #[test]
    fn sem_skills_nao_muda_nada() {
        let m = combat_mods(&[]);
        assert_eq!(m, CombatMods::default());
        let w = weapon_profile("railgun_s").unwrap();
        assert_eq!(apply_to_weapon(&w, &m), w);
    }

    #[test]
    fn no_desconhecido_e_ignorado() {
        // Um cliente que invente ids não pode ganhar nada com isso.
        let m = combat_mods(&["nao_existe".to_string(), "combat_t1".to_string()]);
        assert!((m.damage_mult - 1.05).abs() < 1e-6);
    }

    #[test]
    fn dano_aumenta_com_o_ramo_de_combate() {
        let w = weapon_profile("railgun_s").unwrap();
        let m = combat_mods(&["combat_t1".to_string()]);
        let up = apply_to_weapon(&w, &m);
        assert!(up.damage > w.damage);
        assert!((up.damage - w.damage * 1.05).abs() < 1e-4);
    }

    #[test]
    fn multiplicadores_compoem_por_multiplicacao() {
        // Dois nós de dano: 1.05 * 1.15 = 1.2075, não 1.20.
        let m = combat_mods(&["combat_t1".to_string(), "combat_t3".to_string()]);
        assert!((m.damage_mult - 1.05 * 1.15).abs() < 1e-6);
    }

    #[test]
    fn cadencia_sobe_sem_mexer_no_dano() {
        let w = weapon_profile("plasma_m").unwrap();
        let m = combat_mods(&["combat_t2".to_string()]);
        let up = apply_to_weapon(&w, &m);
        assert!(up.fire_rate > w.fire_rate);
        assert_eq!(up.damage, w.damage);
    }

    #[test]
    fn perfuracao_soma_mas_nao_passa_de_um() {
        let muitos: Vec<String> = std::iter::repeat_n("combat_t4".to_string(), 30).collect();
        let m = combat_mods(&muitos);
        assert!(m.shield_pierce <= 1.0);
        assert!(m.shield_pierce > 0.9);
    }

    #[test]
    fn carga_mais_rapida_preserva_o_multiplicador_de_dano_da_carga() {
        // Reduzir o tempo de carga não pode, por acidente, mudar quanto
        // uma carga cheia rende.
        let w = weapon_profile("lance_singular").unwrap();
        let m = CombatMods {
            charge_time_mult: 0.5,
            ..Default::default()
        };
        let up = apply_to_weapon(&w, &m);
        assert_eq!(up.charge_damage_mult, w.charge_damage_mult);
        assert!((up.charge_time - w.charge_time * 0.5).abs() < 1e-6);
    }

    #[test]
    fn ramos_nao_de_combate_nao_afetam_o_tiro() {
        let m = combat_mods(&["nav_t1".to_string(), "eng_t2".to_string()]);
        assert_eq!(m, CombatMods::default());
    }
}
